src/ctap/crypto_wrapper.rs

```diff
@@ -302,66 +250,30 @@ pub fn encrypt_key_handle(
 /// - the format does not match any known versions,
 /// - the HMAC test fails or
 /// - the relying party does not match the decrypted relying party ID hash.
-///
-/// This functions reads:
-/// - legacy credentials (no version number),
-/// - v1 (ECDSA)
-/// - v2 (EdDSA over curve Ed25519)
+/// - check_cred_protect is set and the credential requires user verification.
 pub fn decrypt_credential_source(
     env: &mut impl Env,
-    credential_id: Vec<u8>,
+    credential_id_bytes: Vec<u8>,
     rp_id_hash: &[u8],
+    check_cred_protect: bool,
 ) -> Result<Option<PublicKeyCredentialSource>, Ctap2StatusCode> {
-    if credential_id.len() < LEGACY_CREDENTIAL_ID_SIZE {
-        return Ok(None);
-    }
-    let hmac_message_size = credential_id.len() - 32;
-    if !verify_hmac_256::<Sha256>(
-        &env.key_store().key_handle_authentication()?,
-        &credential_id[..hmac_message_size],
-        array_ref![credential_id, hmac_message_size, 32],
-    ) {
-        return Ok(None);
-    }
-
-    let (payload, algorithm) = if credential_id.len() == LEGACY_CREDENTIAL_ID_SIZE {
-        (&credential_id[..hmac_message_size], ES256_ALGORITHM)
-    } else {
-        // Version number check
-        let algorithm = match credential_id[0] {
-            ECDSA_CREDENTIAL_ID_VERSION => ES256_ALGORITHM,
-            #[cfg(feature = "ed25519")]
-            ED25519_CREDENTIAL_ID_VERSION => EDDSA_ALGORITHM,
-            _ => return Ok(None),
-        };
-        (&credential_id[1..hmac_message_size], algorithm)
+    let cred_id = match CredentialId::decrypt_from_bytes(env, credential_id_bytes.clone())? {
+        None => return Ok(None),
+        Some(x) => x,
     };
-    if payload.len() != 80 {
-        // We shouldn't have HMAC'ed anything of different length. The check is cheap though.
+    let is_protected =
+        cred_id.cred_protect_policy == Some(CredentialProtectionPolicy::UserVerificationRequired);
+    if rp_id_hash != cred_id.rp_id_hash || (check_cred_protect && is_protected) {
```

Comment:
I don't think this check should be performed here. Logically, it belongs to the GetAssertion flow that should take the returned credential and check its credProtect policy.

Reply:
discoverable credential's check is performed in find_credential, so I perform this here for consistency. It should also make sense to move both checks outside, should I make this change in this PR too?

Reply:
After inspecting the locations I need to change for this, I decided it should be in next PR.

Reply:
Sounds good.